cpp/nccl_comms/src/cuML_comms_nccl_impl.cpp

```diff
@@ -0,0 +1,292 @@
+                return ncclMin;
+            case MLCommon::cumlCommunicator::MAX:
+                return ncclMax;
+        }
+    }
+}
+
+/**
+ * Underlying comms, like NCCL and UCX, should be initialized and ready for use
+ * outside of the cuML Comms lifecycle. This allows us to decouple the ownership
+ * of the actual comms from cuml so that they can also be used directly, outside of
+ * cuml.
+ *
+ * For instance, nccl-py can be used to bootstrap a ncclComm_t before it is
+ * used to construct a cuml comms instance. UCX endpoints can be bootstrapped
+ * in Python as well, before being used to construct a cuML comms instance.
+ */
+void inject_comms(cumlHandle& handle, ncclComm_t comm, ucp_worker_h *ucp_worker, ucp_ep_h **eps, int size, int rank)
+{
+    auto communicator = std::make_shared<MLCommon::cumlCommunicator>(
+         std::unique_ptr<MLCommon::cumlCommunicator_iface>( new cumlNCCLCommunicator_impl(comm, ucp_worker, eps, size, rank) ) );
+    handle.getImpl().setCommunicator( communicator );
+}
+
+cumlNCCLCommunicator_impl::cumlNCCLCommunicator_impl(ncclComm_t comm, ucp_worker_h *ucp_worker, ucp_ep_h **eps, int size, int rank)
+    : _nccl_comm(comm), _ucp_worker(ucp_worker), _ucp_eps(eps), _size(size), _rank(rank), _next_request_id(0) {
+    //initializing NCCL
+//    NCCL_CHECK(ncclCommInitRank(&_nccl_comm, _size, _rank));
+}
+
+cumlNCCLCommunicator_impl::~cumlNCCLCommunicator_impl() {}
+
+int cumlNCCLCommunicator_impl::getSize() const {
+    return _size;
+}
+
+int cumlNCCLCommunicator_impl::getRank() const {
+    return _rank;
+}
+
+std::unique_ptr<MLCommon::cumlCommunicator_iface> cumlNCCLCommunicator_impl::commSplit( int color, int key ) const
+{
+    // Not supported by NCCL
+    printf("commSplit called but not supported in NCCL implementation.\n");
+}
+
+void cumlNCCLCommunicator_impl::barrier() const
+{
+    // @TODO:
+    // 1. Have Rank 0 send out predetermined message and blocks until it gets a response from everyone.
+    // 2. All other ranks block until they see the message and reply with a predetermined response.
+    // 3. Upon getting responses from everyone, Rank 0 sends out a message for everyone to continue.
+    printf("barrier called but not supported in NCCL implementation.\n");
+}
+
+static void send_handle(void *request, ucs_status_t status) {
+    struct ucx_context *context = (struct ucx_context *) request;
+    context->completed = 1;
+}
+
+static void recv_handle(void *request, ucs_status_t status,
+                        ucp_tag_recv_info_t *info) {
+    struct ucx_context *context = (struct ucx_context *) request;
+    context->completed = 1;
+}
+
+
+
+// @TODO: Really, the isend and irecv should be tied to a datatype, as they are in MPI/UCP.
+// This may not matter for p2p, though, if we are just always going to treat them as a contiguous sequence of bytes.
+void cumlNCCLCommunicator_impl::isend(const void *buf, int size, int dest, int tag, request_t *request) const
+{
+  request_t req_id;
```

Comment:
Tagging @Akshay-Venkatesh for feedback. I copied and modified the code from your [ucp_py_ep_send_nb](https://github.com/rapidsai/ucx-py/blob/devel/ucp/_libs/src/ucp_py_ucp_fxns.c#L343) and [ucp_py_recv_nb](https://github.com/rapidsai/ucx-py/blob/devel/ucp/_libs/src/ucp_py_ucp_fxns.c#L248) functions.